File: test-junit5/src/test/java/io/micronaut/test/junit5/JpaNoRollbackTest.java
package io.micronaut.test.junit5;

import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.transaction.TransactionOperations;
import jakarta.inject.Inject;
import jakarta.persistence.EntityManager;
import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaDelete;
import jakarta.persistence.criteria.CriteriaQuery;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

@MicronautTest(rollback = false)
@io.micronaut.context.annotation.Property(name = "datasources.default.url", value = "jdbc:h2:mem:JpaNoRollbackTest;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE")
@DbProperties
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class JpaNoRollbackTest {

    @Inject
    EntityManager entityManager;

    @Inject
    TransactionOperations<?> transactionOperations;

    @AfterAll
    void cleanup() {
        transactionOperations.executeWrite(status -> {
            final CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
            final CriteriaDelete<Book> delete = criteriaBuilder.createCriteriaDelete(Book.class);
            delete.from(Book.class);
            entityManager.createQuery(delete).executeUpdate();
            return null;
        });
    }

    @Test
    void testPersistOne() {
        final Book book = new Book();
        book.setTitle("The Stand");
        entityManager.persist(book);
        assertNotNull(entityManager.find(Book.class, book.getId()));

        final CriteriaQuery<Book> query = entityManager.getCriteriaBuilder().createQuery(Book.class);
        query.from(Book.class);
        assertEquals(1, entityManager.createQuery(query).getResultList().size());
    }

    @Test
    void testPersistTwo() {
        final Book book = new Book();
        book.setTitle("The Shining");
        entityManager.persist(book);
        assertNotNull(entityManager.find(Book.class, book.getId()));

        final CriteriaQuery<Book> query = entityManager.getCriteriaBuilder().createQuery(Book.class);
        query.from(Book.class);
        assertEquals(2, entityManager.createQuery(query).getResultList().size());
    }
}

File: test-kotest5/src/test/kotlin/io/micronaut/test/kotest5/JpaNoRollbackTest.kt
package io.micronaut.test.kotest5

import io.kotest.core.spec.Spec
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest
import io.micronaut.transaction.SynchronousTransactionManager
import io.micronaut.transaction.support.DefaultTransactionDefinition
import jakarta.persistence.EntityManager

@MicronautTest(rollback = false)
@DbProperties
class JpaNoRollbackTest(
    private val entityManager: EntityManager,
    private val transactionManager: SynchronousTransactionManager<Any>
) : BehaviorSpec() {

    init {
        given("no rollback between tests") {
            `when`("test persist one") {
                val book = Book()
                book.title = "The Stand"
                entityManager.persist(book)

                then("the book is persisted") {
                    entityManager.find(Book::class.java, book.id) shouldNotBe null

                    val query = entityManager.criteriaBuilder.createQuery(Book::class.java)
                    query.from(Book::class.java)
                    entityManager.createQuery(query).resultList.size shouldBe 1
                }
            }
        }

        given("a new transaction") {
            `when`("test persist two") {
                val book = Book()
                book.title = "The Shining"
                entityManager.persist(book)

                then("the book is persisted") {
                    entityManager.find(Book::class.java, book.id) shouldNotBe null

                    val query = entityManager.criteriaBuilder.createQuery(Book::class.java)
                    query.from(Book::class.java)
                    entityManager.createQuery(query).resultList.size shouldBe 2
                }
            }
        }
    }

    override suspend fun afterSpec(spec: Spec) {
        val tx = transactionManager.getTransaction(DefaultTransactionDefinition())
        val criteriaBuilder = entityManager.criteriaBuilder
        val delete = criteriaBuilder.createCriteriaDelete(Book::class.java)
        delete.from(Book::class.java)
        entityManager.createQuery(delete).executeUpdate()
        transactionManager.commit(tx)
    }
}
